cmake/genny_subdir.cmake

```diff
@@ -0,0 +1,223 @@
+# Copyright 2019-present MongoDB Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+##
+# Defines the genny_subdir function
+##
+
+#
+# Usage:
+#
+#     genny_subdir(
+#       NAME   project_name
+#       TYPE   SHARED|STATIC|INTERFACE
+#       DEPENDS
+#         list of other libs to link to for output lib
+#       TEST_DEPENDS
+#         list of other libs to link to for test binary
+#       EXECUTABLE name of file to produce for src/main.cpp
+#     )
+#
+# Notes:
+#
+# -   Assumes public headers are in `include` and end with `.hpp`
+# -   Assumes source files are in `src` and end with `.cpp`.
+# -   The file `src/main.cpp` is special. It is not compiled into the
+#     produced library. It is only compiled if the `EXECUTABLE` argument
+#     is given.
+# -   Assumes regular tests are `test/*.cpp` and get parsed by
+#     `ParseAndAddCatchTests`. The test binary is named
+#     `project_name_test`.
+# -   Assumes benchmark tests are `benchmark/*.cpp` and get parsed by
+#     `ParseAndAddCatchTests`. The benchmark test binary is named
+#     `project_name_benchmark`.
+# -   Installs the library to the `GennyLibraryConfig` export
+#
+function(GENNY_SUBDIR)
```

Comment:
I honestly believe that you should have a `add_genny_library` and `add_genny_test` function. This is a standard that tends to come up pretty regularly. They can be almost entirely passthru besides the initial target variable:

```
function(add_genny_library TARGET)
    ...
    add_library(${ARGV})
    ...
endfunction(add_genny_library)

function(add_genny_test TARGET <SOURCES>)
    ...
    add_executable(test_${TARGET} ${SOURCES})
    target_link_library(test_${TARGET} ${TARGET})
    ...
endfunction(add_genny_test)
```

```
add_genny_library(mylib SHARED) # Defines mylib as normal, includes the local includes folder, links in the core deps
add_genny_test(mylib ...) # Defines test_mylib, links it to mylib, adds that silly ParseAndCombulate script
```

Reply:
> I honestly believe that you should have a add_genny_library and add_genny_test function...

Not quite sure I see the benefit - What does this get you apart from the ability to specify multiple distinct test targets? why would we want multiple of them beyond just the "regular" ones and "benchmark" ones?

Reply:
Well, I can think of a few different short term goals that this would thwart:
1. You no longer have the direct ability to bundle code dependent upon features into a static library that gets added if needed. (Ala what was proposed with jasperlib)
2. You no longer get the ability separate long running tests from short running tests or test that need specific features from test that are vanilla unit tests.
3. You now have a weight of cmake function to support for every bit of your linking and you'll slowly have to modify it to fit your needs.